src/sender_queue/mod.rs

```diff
@@ -0,0 +1,366 @@
+    outgoing_queue: OutgoingQueue<D>,
+    /// Known current epochs of remote nodes.
+    peer_epochs: BTreeMap<D::NodeId, <D::Message as Epoched>::Epoch>,
+}
+
+pub type Step<D> = ::Step<SenderQueue<D>>;
+
+pub type Result<T, D> = ::std::result::Result<T, <D as DistAlgorithm>::Error>;
+
+impl<D> DistAlgorithm for SenderQueue<D>
+where
+    D: DistAlgorithm + Debug + Send + Sync + SenderQueueFunc<D>,
+    D::Message: Clone + Epoched + Serialize + DeserializeOwned,
+    D::NodeId: NodeIdT + Rand,
+{
+    type NodeId = D::NodeId;
+    type Input = D::Input;
+    type Output = D::Output;
+    type Message = Message<D::Message>;
+    type Error = D::Error;
+
+    fn handle_input(&mut self, input: Self::Input) -> Result<Step<D>, D> {
+        let mut step = self.algo.handle_input(input)?;
+        let mut sender_queue_step = self.update_epoch(&step);
+        self.defer_messages(&mut step);
+        sender_queue_step.extend(step.map(|output| output, Message::from));
+        Ok(sender_queue_step)
+    }
+
+    fn handle_message(
+        &mut self,
+        sender_id: &D::NodeId,
+        message: Self::Message,
+    ) -> Result<Step<D>, D> {
+        match message.content {
+            MessageContent::EpochStarted => Ok(self.handle_epoch_started(sender_id, message.epoch)),
+            MessageContent::Algo(msg) => self.handle_message_content(sender_id, msg),
+        }
+    }
+
+    fn terminated(&self) -> bool {
+        false
+    }
+
+    fn our_id(&self) -> &D::NodeId {
+        &self.our_id
+    }
+}
+
+impl<D> SenderQueue<D>
+where
+    D: DistAlgorithm + Debug + Send + Sync + SenderQueueFunc<D>,
+    D::Message: Clone + Epoched + Serialize + DeserializeOwned,
+    D::NodeId: NodeIdT + Rand,
+{
+    /// Returns a new `SenderQueueBuilder` configured to manage a given `DynamicHoneyBadger` instance.
+    pub fn builder(algo: D) -> SenderQueueBuilder<D> {
+        SenderQueueBuilder::new(algo)
+    }
+
+    /// Handles an epoch start announcement.
+    fn handle_epoch_started(
+        &mut self,
+        sender_id: &D::NodeId,
+        epoch: <D::Message as Epoched>::Epoch,
+    ) -> Step<D> {
+        self.peer_epochs
+            .entry(sender_id.clone())
+            .and_modify(|e| {
+                if *e < epoch {
+                    *e = epoch;
+                }
+            }).or_insert(epoch);
+        self.remove_earlier_messages(sender_id, epoch);
+        self.process_new_epoch(sender_id, epoch)
+    }
+
+    /// Removes all messages queued for the remote node from epochs upto `epoch`.
+    fn remove_earlier_messages(
+        &mut self,
+        sender_id: &D::NodeId,
+        epoch: <D::Message as Epoched>::Epoch,
+    ) {
+        let earlier_keys: Vec<_> = self
+            .outgoing_queue
+            .keys()
+            .cloned()
+            .filter(|(id, this_epoch)| {
+                id == sender_id
+                    && PartialOrd::partial_cmp(this_epoch, &epoch) == Some(Ordering::Less)
```

Comment:
Isn't that equivalent to `this_epoch < epoch`? I think `<` doesn't evaluate to `true` if `partial_cmp` returns `None`?